src/bvals/cc/sparse_bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,393 @@
+    if (buf.GetState() == BufferState::received && !v->IsAllocated()) {
+      pmb->AllocateSparse(v->label());
+      // TODO: Need to flag this so that the array gets filled with something sensible
+    }
+  });
+  
+  Kokkos::Profiling::popRegion();
+
+  if (all_received) return TaskStatus::complete;
+  return TaskStatus::incomplete;
+
+}
+
+TaskStatus SetInternalSparseBoundaryBuffers(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_SetInternalBoundaries");  
+  
+  Mesh* pmesh = md->GetMeshPointer();
+  
+  // Check for rebuild 
+  bool rebuild = false;
+  int nbound = 0;
+  {
+  WriteRegion region("check for rebuild receive"); 
+  IterateBoundaries(md, [&](sp_mb_t pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
+    auto& buf = pmesh->boundary_comm_map[{nb.snb.gid, pmb->gid, v->label()}];
+    if (nbound < md->recv_bnd_info_h.size()) {    
+      rebuild = rebuild || !UsingSameResource(md->recv_bnd_info_h(nbound).buf, buf);
+      if ((buf.GetState() == BufferState::received) && !md->recv_bnd_info_h(nbound).allocated) rebuild = true;
+      if ((buf.GetState() == BufferState::received_null) && md->recv_bnd_info_h(nbound).allocated) rebuild = true;
+    } else { 
+      rebuild = true; 
+    }
+    ++nbound;
+  });
+  }
+  if (rebuild) {
+    WriteRegion region("rebuild receive"); 
+    md->recv_bnd_info = BufferCache_t("recv_boundary_info", nbound);
+    md->recv_bnd_info_h = Kokkos::create_mirror_view(md->recv_bnd_info); 
+    int iarr = 0;
+    IterateBoundaries(md, [&](sp_mb_t pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
+      auto& buf = pmesh->boundary_comm_map[{nb.snb.gid, pmb->gid, v->label()}];
+      
+      md->recv_bnd_info_h(iarr) = BndInfo::Setter(pmb, nb, v); 
+      md->recv_bnd_info_h(iarr).buf = buf; 
+      if (buf.GetState() == BufferState::received) {
+        md->recv_bnd_info_h(iarr).allocated = true;
+      } 
+      else if (buf.GetState() == BufferState::received_null) { 
+        md->recv_bnd_info_h(iarr).allocated = false;
+      } else {
+        PARTHENON_FAIL("Buffer should be in received state.");
+      }
+
+      ++iarr;
+    }); 
+    Kokkos::deep_copy(md->recv_bnd_info, md->recv_bnd_info_h); 
+  }
+  
+  //const Real threshold = Globals::sparse_config.allocation_threshold;
+  auto& bnd_info = md->recv_bnd_info;
+  {
+  WriteRegion region("set boundary buffers"); 
+  Kokkos::parallel_for(
+    "SetBoundaryBuffers",
+    Kokkos::TeamPolicy<>(parthenon::DevExecSpace(), nbound, Kokkos::AUTO),
+    KOKKOS_LAMBDA(parthenon::team_mbr_t team_member) {
+      const int b = team_member.league_rank();
+      
+      const int &si = bnd_info(b).si;
+      const int &ei = bnd_info(b).ei;
+      const int &sj = bnd_info(b).sj;
+      const int &ej = bnd_info(b).ej;
+      const int &sk = bnd_info(b).sk;
+      const int &ek = bnd_info(b).ek;
+
+      const int Ni = ei + 1 - si;
+      const int Nj = ej + 1 - sj;
+      const int Nk = ek + 1 - sk;
+      const int &Nv = bnd_info(b).Nv;
```

Comment:
Another `Nt`, `Nu` place (just calling these out in case you want to add comments for later)